File: qbopt/backend/farload.py
"""Select complete far-pointer loads before allocation loses their address.

A far pointer reaches lowering as two word loads -- its offset, then its
selector -- whether a frontend spelt it so or `narrow` split a dword read only
through its halves.  They are one far load when the high word is only ever a
cell's selector; that use, not a source type, is what makes `les` the right
instruction.  The final physical peephole can
recognise that shape only when both loads still use the same physical address
register.  Under pressure the allocator rematerializes a near owner for each
word, so it is too late there even though the pre-allocation LIR still proves
the words are adjacent parts of one cell.

This is lowering, not an LIR optimization pass.  It chooses the one x86 form
for a machine-neutral pair of loads while the pair's address identity is
available; allocation remains responsible for giving the offset an address
register and the selector ES, FS or GS.
"""

from dataclasses import replace

from qbopt.model import ir
from qbopt.model import lir


def selected(insns: tuple[lir.Insn, ...], selectors: frozenset[int]) -> tuple[lir.Insn, ...]:
    """Select private ``mov offset,[p]; mov selector,[p+2]`` pairs as ``les``.

    A combined load reads both words before writing either result, unlike the
    two-source sequence.  Refuse a pair if either result reaches either
    address; in that case the two forms can observe different addresses.
    Source-backed operations are likewise left to the post-allocation
    peephole, which owns the necessary byte-coverage rewrite.  Fresh C OMF
    operations carry zero-width source anchors, which remain safe to join.
    """
    made: dict[int, lir.Insn] = {}
    erased: set[int] = set()
    for at, first in enumerate(insns[:-1]):
        if at in erased:
            continue
        second = insns[at + 1]
        joined = _pair(first, second, selectors)
        if joined is None:
            continue
        made[at] = joined
        erased.add(at + 1)
    return tuple(
        made[at] if at in made else replace(lir.anchor(one), defines=(), uses=(), widths=()) if at in erased else one
        for at, one in enumerate(insns)
    )


def _pair(first: lir.Insn, second: lir.Insn, selectors: frozenset[int]) -> lir.Insn | None:
    if (
        not _plain(first)
        or not _plain(second)
        or first.covers is not None
        and first.covers[0] != first.covers[1]
        or second.covers is not None
        and second.covers[0] != second.covers[1]
    ):
        return None
    words = []
    for one in (first, second):
        match one.what:
            case ir.Semantics(ir.Operation.MOVE, "mov", (ir.Held() as dest,), (ir.Mem() as cell,)):
                if dest.width != cell.width != 2:
                    return None
                words.append((dest, cell))
            case _:
                return None
    (first_dest, first_cell), (second_dest, second_cell) = words
    if not _next_word(first_cell, second_cell) or second_dest.value not in selectors or _volatile(first, second):
        return None
    # A fixed address survives allocation unchanged.  Leave its two semantic
    # values independent so spilling may rematerialize either word from the
    # original cell and selector allocation may choose ES, FS or GS directly;
    # the physical far-load peephole can still fuse adjacent survivors.  A
    # virtual base/selector/index is different: allocation can reconstruct its
    # owner separately for each word, destroying the common address before the
    # physical peephole sees it.  Only that genuinely lossy boundary needs the
    # early complete-load selection performed here.
    if not tuple(ir.values(first_cell)):
        return None
    defined = {first_dest.value, second_dest.value}
    if any(value.value in defined for cell in (first_cell, second_cell) for value in ir.values(cell)):
        return None
    # Spelt `les`; the rewriter respells it for the segment register the
    # selector is given.
    what = ir.Semantics(ir.Operation.MOVE, "les", (first_dest, second_dest), (replace(first_cell, width=4),))
    return replace(
        first,
        what=what,
        defines=tuple(dict.fromkeys((*first.defines, *second.defines))),
        uses=tuple(dict.fromkeys((*first.uses, *second.uses))),
    )


def _plain(one: lir.Insn) -> bool:
    return not (
        one.clobbers
        or one.clobbers_high
        or one.requires
        or one.delivers
        or one.spread
        or one.group is not None
        or one.symbol is True
        or one.frame_adjust
        or one.spill_reload
        or one.spill_store
        or one.rematerialized
    )


def _next_word(low: ir.Mem, high: ir.Mem) -> bool:
    """Whether ``high`` is the word immediately after ``low`` by one address."""
    same = lambda cell: replace(cell, addr=None if cell.addr is None else replace(cell.addr, disp=0), offset=0)  # noqa: E731
    if same(low) != same(high):
        return False
    if low.addr is None or high.addr is None:
        return low.addr is None and high.addr is None and high.offset == low.offset + 2
    moved = high.addr.disp - low.addr.disp
    return moved == 2 and high.offset - low.offset in (0, 2) or moved == 0 and high.offset == low.offset + 2


def _volatile(first: lir.Insn, second: lir.Insn) -> bool:
    return any(ref.volatile for one in (first, second) if one.op is not None for ref in one.op.loads)


def selectors(made: dict[int, tuple[lir.Insn, ...]], kept: frozenset[int]) -> frozenset[int]:
    """Values read only as a cell's selector: what makes a far load the right load.

    `kept` holds values a phi reads, which no instruction here shows.
    """
    selecting: set[int] = set()
    numeric: set[int] = set(kept)
    for block in made.values():
        for one in block:
            numeric.update(held.value for held, _ in (*one.requires, *one.delivers))
            if one.what is None:
                continue
            for where in (*one.what.dests, *one.what.sources):
                if isinstance(where, ir.Mem):
                    numeric.update(held.value for held in (where.base, where.index) if held is not None)
                    if where.selector is not None:
                        selecting.add(where.selector.value)
            numeric.update(where.value for where in one.what.sources if isinstance(where, ir.Held))
    return frozenset(selecting - numeric)

File: src/backend/farload.rs
//! Port of `qbopt/backend/farload.py`: select complete far-pointer loads
//! before allocation loses their address.

use std::collections::BTreeSet;
use std::sync::Arc;

use indexmap::{IndexMap, IndexSet};

use crate::model::ir::{self, Held, Loc, Mem, Operation, Semantics};
use crate::model::lir::{self, Insn};

/// Select private `mov offset,[p]; mov selector,[p+2]` pairs as `les`.
pub fn selected(insns: &[Arc<Insn>], selectors: &BTreeSet<u32>) -> Vec<Arc<Insn>> {
    let mut made: IndexMap<usize, Arc<Insn>> = IndexMap::new();
    let mut erased: BTreeSet<usize> = BTreeSet::new();
    for (at, first) in insns[..insns.len().saturating_sub(1)].iter().enumerate() {
        if erased.contains(&at) {
            continue;
        }
        let second = &insns[at + 1];
        let Some(joined) = _pair(first, second, selectors) else {
            continue;
        };
        made.insert(at, joined);
        erased.insert(at + 1);
    }
    insns
        .iter()
        .enumerate()
        .map(|(at, one)| {
            if let Some(joined) = made.get(&at) {
                Arc::clone(joined)
            } else if erased.contains(&at) {
                let mut anchored = (*lir::anchor(Arc::clone(one))).clone();
                anchored.defines = Vec::new();
                anchored.uses = Vec::new();
                anchored.widths = Vec::new();
                Arc::new(anchored)
            } else {
                Arc::clone(one)
            }
        })
        .collect()
}

fn _pair(first: &Insn, second: &Insn, selectors: &BTreeSet<u32>) -> Option<Arc<Insn>> {
    if !_plain(first)
        || !_plain(second)
        || first.covers.is_some_and(|covers| covers.0 != covers.1)
        || second.covers.is_some_and(|covers| covers.0 != covers.1)
    {
        return None;
    }
    let mut words: Vec<(Held, Mem)> = Vec::new();
    for one in [first, second] {
        match &one.what {
            Some(what)
                if what.op == Operation::Move
                    && what.name.as_deref() == Some("mov")
                    && matches!(what.dests.as_slice(), [Loc::Held(_)])
                    && matches!(what.sources.as_slice(), [Loc::Mem(_)]) =>
            {
                let (Loc::Held(dest), Loc::Mem(cell)) = (&what.dests[0], &what.sources[0]) else {
                    unreachable!()
                };
                // Python's chained `dest.width != cell.width != 2`.
                if dest.width != cell.width && cell.width != 2 {
                    return None;
                }
                words.push((*dest, cell.clone()));
            }
            _ => return None,
        }
    }
    let (first_dest, first_cell) = words[0].clone();
    let (second_dest, second_cell) = words[1].clone();
    if !_next_word(&first_cell, &second_cell) || !selectors.contains(&second_dest.value) || _volatile(first, second) {
        return None;
    }
    // A fixed address survives allocation unchanged; only a virtual
    // base/selector/index needs the early complete-load selection.
    if ir::values(&Loc::Mem(first_cell.clone())).is_empty() {
        return None;
    }
    let defined: BTreeSet<u32> = [first_dest.value, second_dest.value].into_iter().collect();
    if [&first_cell, &second_cell]
        .into_iter()
        .any(|cell| ir::values(&Loc::Mem(cell.clone())).iter().any(|value| defined.contains(&value.value)))
    {
        return None;
    }
    // Spelt `les`; the rewriter respells it for the segment register the
    // selector is given.
    let what = Semantics {
        name: Some("les".to_owned()),
        dests: vec![Loc::Held(first_dest), Loc::Held(second_dest)],
        sources: vec![Loc::Mem(Mem { width: 4, ..first_cell })],
        ..Semantics::new(Operation::Move)
    };
    let mut joined = first.clone();
    joined.what = Some(what);
    joined.defines = first
        .defines
        .iter()
        .chain(&second.defines)
        .copied()
        .collect::<IndexSet<u32>>()
        .into_iter()
        .collect();
    joined.uses = first
        .uses
        .iter()
        .chain(&second.uses)
        .copied()
        .collect::<IndexSet<u32>>()
        .into_iter()
        .collect();
    Some(Arc::new(joined))
}

fn _plain(one: &Insn) -> bool {
    !(!one.clobbers.is_empty()
        || !one.clobbers_high.is_empty()
        || !one.requires.is_empty()
        || !one.delivers.is_empty()
        || !one.spread.is_empty()
        || one.group.is_some()
        || one.symbol == Some(true)
        || one.frame_adjust
        || one.spill_reload
        || one.spill_store
        || one.rematerialized)
}

/// Whether `high` is the word immediately after `low` by one address.
fn _next_word(low: &Mem, high: &Mem) -> bool {
    let same = |cell: &Mem| Mem {
        addr: cell.addr.map(|addr| ir::Addr { disp: 0, ..addr }),
        offset: 0,
        ..cell.clone()
    };
    if same(low) != same(high) {
        return false;
    }
    let (Some(low_addr), Some(high_addr)) = (low.addr, high.addr) else {
        return low.addr.is_none() && high.addr.is_none() && high.offset == low.offset + 2;
    };
    let moved = high_addr.disp - low_addr.disp;
    moved == 2 && [0, 2].contains(&(high.offset - low.offset)) || moved == 0 && high.offset == low.offset + 2
}

fn _volatile(first: &Insn, second: &Insn) -> bool {
    [first, second]
        .into_iter()
        .filter_map(|one| one.op.as_ref())
        .any(|op| op.loads.iter().any(|r#ref| r#ref.volatile))
}

/// Values read only as a cell's selector: what makes a far load the right load.
///
/// `kept` holds values a phi reads, which no instruction here shows.
pub fn selectors(made: &IndexMap<i64, Vec<Arc<Insn>>>, kept: &BTreeSet<u32>) -> BTreeSet<u32> {
    let mut selecting: BTreeSet<u32> = BTreeSet::new();
    let mut numeric: BTreeSet<u32> = kept.clone();
    for block in made.values() {
        for one in block {
            numeric.extend(one.requires.iter().chain(&one.delivers).map(|(held, _)| held.value));
            let Some(what) = &one.what else {
                continue;
            };
            for r#where in what.dests.iter().chain(&what.sources) {
                if let Loc::Mem(cell) = r#where {
                    numeric.extend([cell.base, cell.index].into_iter().flatten().map(|held| held.value));
                    if let Some(selector) = cell.selector {
                        selecting.insert(selector.value);
                    }
                }
            }
            numeric.extend(what.sources.iter().filter_map(|r#where| match r#where {
                Loc::Held(held) => Some(held.value),
                _ => None,
            }));
        }
    }
    selecting.difference(&numeric).copied().collect()
}

#[cfg(test)]
mod tests {
    //! Port of `tests/test_farload.py`.

    use std::collections::BTreeSet;
    use std::sync::Arc;

    use super::selected;
    use crate::model::ir::{self, Addr, Loc, Mem, Operation, Semantics, Space};
    use crate::model::lir::Insn;
    use crate::model::mir::{self, Arg, Cell, Kind, MemRef, Op, OpCode, Value};

    fn typed_ref(space: Space, disp: i64) -> MemRef {
        MemRef { typed: Some(("pointer4".to_owned(), false)), ..MemRef::new(Some(Addr::new(space, disp)), 2) }
    }

    fn load(at: i64, value: Value, r#ref: MemRef, cell: Mem, uses: Vec<u32>) -> Arc<Insn> {
        let destination = ir::Held { value: value.id, width: 2 };
        let mut op = Op::new(at, Some(OpCode::Operation(Operation::Move)), "mov", vec![value], Vec::new());
        op.loads = vec![r#ref.clone()];
        op.kind = Kind::Load;
        op.args = vec![Arg::Cell(Cell { r#ref })];
        op.results = vec![Arg::Held(mir::Held { value, width: 2 })];
        let mut one = Insn::new(
            at,
            Some((at, at)),
            Some(Semantics {
                name: Some("mov".to_owned()),
                dests: vec![Loc::Held(destination)],
                sources: vec![Loc::Mem(cell)],
                ..Semantics::new(Operation::Move)
            }),
            vec![value.id],
            uses,
        );
        one.op = Some(Arc::new(op));
        Arc::new(one)
    }

    #[test]
    fn test_adjacent_words_are_one_far_load_only_when_the_high_word_is_a_selector() {
        // qcport's dynamic far-struct fields emitted two loads per far pointer.
        let (first_value, second_value) = (Value::new(1, 1), Value::new(2, 1));
        let first_ref = typed_ref(Space::Far, 4);
        let second_ref = typed_ref(Space::Far, 6);
        let cell = |r#ref: &MemRef| Mem {
            base: Some(ir::Held { value: 10, width: 2 }),
            selector: Some(ir::Held { value: 11, width: 2 }),
            ..Mem::new(r#ref.addr, 2)
        };
        let values = |cell: &Mem| ir::values(&Loc::Mem(cell.clone())).iter().map(|one| one.value).collect();
        let (first_cell, second_cell) = (cell(&first_ref), cell(&second_ref));
        let original = vec![
            load(1, first_value, first_ref, first_cell.clone(), values(&first_cell)),
            load(2, second_value, second_ref, second_cell.clone(), values(&second_cell)),
        ];

        let selected = selected(&original, &BTreeSet::from([2]));

        assert_eq!(selected[0].what.as_ref().unwrap().name.as_deref(), Some("les"));
        let Loc::Mem(source) = &selected[0].what.as_ref().unwrap().sources[0] else { panic!() };
        assert_eq!(source.width, 4);
        assert_eq!(selected[0].defines, vec![1, 2]);
        assert_eq!(selected[1].what.as_ref().unwrap().op, Operation::Nothing);
        // Adjacent words whose high one is a number, not a selector, stay two loads.
        assert_eq!(super::selected(&original, &BTreeSet::new()), original);
    }

    #[test]
    fn test_fixed_far_pointer_load_defers_fusion_until_after_allocation() {
        // indexed.lru_use's fixed parameter pair became an eager LES.
        let values = [Value::new(1, 1), Value::new(2, 1)];
        let load = |at: i64, value: Value, displacement: i64| {
            let r#ref = typed_ref(Space::Frame, displacement);
            let cell = Mem::new(r#ref.addr, 2);
            load(at, value, r#ref, cell, Vec::new())
        };

        let original = vec![load(1, values[0], 4), load(2, values[1], 6)];

        assert_eq!(selected(&original, &BTreeSet::from([2])), original);
    }
}
